src/native/minipal/memorybarrierprocesswide.c

```diff
@@ -0,0 +1,232 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef HOST_WINDOWS
+#include <assert.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <minipal/memorybarrierprocesswide.h>
+
+#ifdef __APPLE__
+#include <stdlib.h>
+#include <mach/thread_state.h>
+
+#define CHECK_MACH(_msg, machret) do {                                      \
+        if (machret != KERN_SUCCESS)                                        \
+        {                                                                   \
+            char _szError[1024];                                            \
+            snprintf(_szError, ARRAY_SIZE(_szError), "%s: %u: %s", __FUNCTION__, __LINE__, _msg);  \
+            mach_error(_szError, machret);                                  \
+            abort();                                                        \
+        }                                                                   \
+    } while (false)
+
+#endif // __APPLE__
+
+#ifdef __linux__
+#include <linux/membarrier.h>
+#include <sys/syscall.h>
+#define membarrier(...) syscall(__NR_membarrier, __VA_ARGS__)
+#elif HAVE_SYS_MEMBARRIER_H
+#include <sys/membarrier.h>
+#ifdef TARGET_BROWSER
+#define membarrier(cmd, flags, cpu_id) 0 // browser/wasm is currently single threaded
+#endif
+#endif
+
+static bool CanFlushUsingMembarrier(void)
+{
+#if defined(__linux__) || HAVE_SYS_MEMBARRIER_H
+
+#ifdef TARGET_ANDROID
+    // Avoid calling membarrier on older Android versions where membarrier
+    // may be barred by seccomp causing the process to be killed.
+    int apiLevel = android_get_device_api_level();
+    if (apiLevel < __ANDROID_API_Q__)
+    {
+        return false;
+    }
+#endif
+
+    // Starting with Linux kernel 4.14, process memory barriers can be generated
+    // using MEMBARRIER_CMD_PRIVATE_EXPEDITED.
+
+    int mask = membarrier(MEMBARRIER_CMD_QUERY, 0, 0);
+
+    if (mask >= 0 &&
+        mask & MEMBARRIER_CMD_PRIVATE_EXPEDITED &&
+        // Register intent to use the private expedited command.
+        membarrier(MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED, 0, 0) == 0)
+    {
+        return true;
+    }
+#endif
+
+    return false;
+}
+
+//
+// Tracks if the OS supports FlushProcessWriteBuffers using membarrier
+//
+static int s_flushUsingMemBarrier = 0;
+
+#ifndef TARGET_APPLE
+// Helper memory page used by the FlushProcessWriteBuffers
+static uint8_t* g_helperPage = 0;
+
+// Mutex to make the FlushProcessWriteBuffersMutex thread safe
+static pthread_mutex_t g_flushProcessWriteBuffersMutex;
+
+static size_t s_pageSize = 0;
+#endif // !TARGET_APPLE
+
+bool minipal_initialize_memory_barrier_process_wide(void)
+{
+#ifndef TARGET_WASM
```

Comment:
```suggestion
#ifdef HOST_WASM
    // browser/wasm is currently single threaded
#elif defined(HOST_APPLE)
...

#else
#if HAVE_SYS_MEMBARRIER_H
    if (CanFlushUsingMembarrier())
    {
        s_flushUsingMemBarrier = true;
    }
    else
#else
    {
        // Fallback implementation
    }
#endif
```
I think conditions with positive polarity would make the code more readable.

Same structure for minipal_memory_barrier_process_wide